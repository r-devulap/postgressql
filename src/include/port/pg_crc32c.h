/*-------------------------------------------------------------------------
 *
 * pg_crc32c.h
 *	  Routines for computing CRC-32C checksums.
 *
 * The speed of CRC-32C calculation has a big impact on performance, so we
 * jump through some hoops to get the best implementation for each
 * platform. Some CPU architectures have special instructions for speeding
 * up CRC calculations (e.g. Intel SSE 4.2), on other platforms we use the
 * Slicing-by-8 algorithm which uses lookup tables.
 *
 * The public interface consists of four macros:
 *
 * INIT_CRC32C(crc)
 *		Initialize a CRC accumulator
 *
 * COMP_CRC32C(crc, data, len)
 *		Accumulate some (more) bytes into a CRC
 *
 * FIN_CRC32C(crc)
 *		Finish a CRC calculation
 *
 * EQ_CRC32C(c1, c2)
 *		Check for equality of two CRCs.
 *
 * Portions Copyright (c) 1996-2025, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/port/pg_crc32c.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef PG_CRC32C_H
#define PG_CRC32C_H

#include "pg_cpu.h"
#include "port/pg_bswap.h"

typedef uint32 pg_crc32c;

/* The INIT and EQ macros are the same for all implementations. */
#define INIT_CRC32C(crc) ((crc) = 0xFFFFFFFF)
#define EQ_CRC32C(c1, c2) ((c1) == (c2))
#define FIN_CRC32C(crc) ((crc) ^= 0xFFFFFFFF)

/* x86 / x86_64 */
#if defined(PG_CPU_X86) || defined(PG_CPU_x86_64)
extern pg_crc32c pg_comp_crc32c_sb8(pg_crc32c crc, const void *data, size_t len);
extern pg_crc32c pg_comp_crc32c_sse42(pg_crc32c crc, const void *data, size_t len);
extern pg_crc32c pg_comp_crc32c_pclmul(pg_crc32c crc, const void *data, size_t len);
extern pg_crc32c (*pg_comp_crc32c) (pg_crc32c crc, const void *data, size_t len);
#define COMP_CRC32C(crc, data, len) \
	((crc) = pg_comp_crc32c((crc), (data), (len)))

/* ARMV8 */
#elif defined(USE_ARMV8_CRC32C)
extern pg_crc32c pg_comp_crc32c_armv8(pg_crc32c crc, const void *data, size_t len);
#define COMP_CRC32C(crc, data, len)							\
	((crc) = pg_comp_crc32c_armv8((crc), (data), (len)))

/* ARMV8 with runtime check */
#elif defined(USE_ARMV8_CRC32C_WITH_RUNTIME_CHECK)
extern pg_crc32c pg_comp_crc32c_sb8(pg_crc32c crc, const void *data, size_t len);
extern pg_crc32c pg_comp_crc32c_armv8(pg_crc32c crc, const void *data, size_t len);
extern pg_crc32c (*pg_comp_crc32c) (pg_crc32c crc, const void *data, size_t len);
#define COMP_CRC32C(crc, data, len) \
	((crc) = pg_comp_crc32c((crc), (data), (len)))

/* LoongArch */
#elif defined(USE_LOONGARCH_CRC32C)
extern pg_crc32c pg_comp_crc32c_loongarch(pg_crc32c crc, const void *data, size_t len);
#define COMP_CRC32C(crc, data, len)							\
	((crc) = pg_comp_crc32c_loongarch((crc), (data), (len)))

#else
/*
 * Use slicing-by-8 algorithm.
 *
 * On big-endian systems, the intermediate value is kept in reverse byte
 * order, to avoid byte-swapping during the calculation. FIN_CRC32C reverses
 * the bytes to the final order.
 */
#define COMP_CRC32C(crc, data, len) \
	((crc) = pg_comp_crc32c_sb8((crc), (data), (len)))
#ifdef WORDS_BIGENDIAN
#undef FIN_CRC32C
#define FIN_CRC32C(crc) ((crc) = pg_bswap32(crc) ^ 0xFFFFFFFF)
#endif

extern pg_crc32c pg_comp_crc32c_sb8(pg_crc32c crc, const void *data, size_t len);
#endif

#endif							/* PG_CRC32C_H */
