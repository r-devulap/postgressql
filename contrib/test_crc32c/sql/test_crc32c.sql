CREATE EXTENSION test_crc32c;

select drive_crc32c(1, i) from generate_series(100, 300, 4) i;
