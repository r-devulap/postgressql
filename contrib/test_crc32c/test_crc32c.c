/* select drive_crc32c(1000000, 1024); */

#include "postgres.h"
#include "fmgr.h"
#include "port/pg_crc32c.h"
#include "common/pg_prng.h"

PG_MODULE_MAGIC;

/*
 * drive_crc32c(count: int, num: int) returns bigint
 *
 * count is the nuimber of loops to perform
 *
 * num is the number byte in the buffer to calculate
 * crc32c over.
 */
PG_FUNCTION_INFO_V1(drive_crc32c);
Datum
drive_crc32c(PG_FUNCTION_ARGS)
{
	int64			count	= PG_GETARG_INT64(0);
	int64			num		= PG_GETARG_INT64(1);
	char*		data	= malloc((size_t)num);
	pg_crc32c crc;
	pg_prng_state state;
	uint64 seed = 42;
	pg_prng_seed(&state, seed);
	/* set random data */
	for (uint64 i = 0; i < num; i++)
	{
		data[i] = pg_prng_uint32(&state) % 255;
	}

	INIT_CRC32C(crc);

	while(count--)
	{
		INIT_CRC32C(crc);
		COMP_CRC32C(crc, data, num);
		FIN_CRC32C(crc);
	}

	free((void *)data);

	PG_RETURN_INT64((int64_t)crc);
}
