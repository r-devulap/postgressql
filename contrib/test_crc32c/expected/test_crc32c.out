CREATE EXTENSION test_crc32c;
select drive_crc32c(1, i) from generate_series(100, 300, 4) i;
 drive_crc32c 
--------------
    532139994
   2103623867
    785984197
   2686825890
   3213049059
   3819630168
   1389234603
    534072900
   2930108140
   2496889855
   1475239611
    136366931
   3067402116
   2012717871
   3682416023
   2054270645
   1817339875
   4100939569
   1192727539
   3636976218
    369764421
   3161609879
   1067984880
   1235066769
   3138425899
    648132037
   4203750233
   1330187888
   2683521348
   1951644495
   2574090107
   3904902018
   3772697795
   1644686344
   2868962106
   3369218491
   3902689890
   3456411865
    141004025
   1504497996
   3782655204
   3544797610
   3429174879
   2524728016
   3935861181
     25498897
    692684159
    345705535
   2761600287
   2654632420
   3945991399
(51 rows)

